### DISCUSSION TITLE
Site Navbar Fetching Data Based on URL

### DISCUSSION THREAD

**Discussion by didomenicom** [vercel/next.js#51022] [Help]
### Summary

I am trying to build an application that has a navbar defined in the [RootLayout file](https://github.com/didomenicom/nextjs13-navbar-fetch-example/blob/main/src/app/layout.tsx#L16) and will load content based on the URL. [Here](https://github.com/didomenicom/nextjs13-navbar-fetch-example) is a very simple version of what I am trying to accomplish. 

For example, if a user goes to /users/<id>, it will load their [user profile](https://github.com/didomenicom/nextjs13-navbar-fetch-example/blob/main/src/app/user/%5BuserID%5D/page.tsx#L29) (details not implemented, but this is a server component and the request to fetch data is happening server side). At the same time, the navbar should display the user's name if they are on the user profile page. 

The only way I have been able to get this working is creating a [NavbarUser](https://github.com/didomenicom/nextjs13-navbar-fetch-example/blob/main/src/components/NavbarUser.tsx#L9) client component that fetches the user's data client side. Ideally, this would be a server component as well so it can make use of the already cached request to the API that was done loading the profile page. 

From what I have found online [[1](https://stackoverflow.com/questions/75861004/accessing-search-params-from-layout-or-component-file-in-next-js-13), [2](https://stackoverflow.com/questions/75741928/how-to-get-params-in-a-client-component-in-next13)], it seems like there is no way to get the URL params in the RootLayout or Navbar without those components being a client component. 

I think it could be possible to create a client component that grabs the userID and passes it to a server component. Another option could be to create an API route that handles the client request to mask auth keys. Both options I can think of seem incorrect. 

What do you recommend as the best approach for fetching URL specific data in the navbar, and can (or should) it be done as a server component? 

### Example

https://github.com/didomenicom/nextjs13-navbar-fetch-example

**Comment by devjiwonchoi** [comment]
Your approach seems to be great, Next.js `App Router` does not provide getting the URL parameter as `Server Component` by default.
One piece of advice is that you can use the `useParams` instead of going around with `usePathname`.

Here's a modified code of your `NavbarUser.tsx`:

```js
// NavbarUser.tsx
'use client';

import Link from 'next/link';
import { useParams } from 'next/navigation';
import { User, getUser } from '@/lib/getUser';

type Props = {};

export default async function NavbarUser({}: Props) {
  const { userID } = useParams();
  const user: User | undefined = await getUser(`${userID}`);
  
  return user ? <Link href={`/user/${user?.id}`}>{user.name}</Link> : <Link href="/user/2">My Profile</Link>;
}
```

## Modifications

1. Removed `import React from 'react'` since it is not required anymore since `React v.17`.
> With the new transform, you can use JSX without importing React.
2. Used `useParams` to get the param values instead of `usePathname` for to meet the purpose.
> useParams is a Client Component hook that lets you read a route's [dynamic params](https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes) filled in by the current URL.
3. Destructured return object of `useParams` as `{ userID }` in order to get the value of userID as variable `userID`.
4. Used ternary operator for displaying the `Link`.

## Conclusion
Your approach is great, but use `useParams` instead of `usePathname`.

## Reference
- [Introducing the New JSX Transform](https://legacy.reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html)
- [Next.js App Router `useParams`](https://nextjs.org/docs/app/api-reference/functions/use-params)
- [Destructuring Assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment)
- [Ternary Operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_operator)

**Comment by icyJoseph** [comment]
I have a way, that theoretically should work, but this https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#client-side-caching-of-rendered-server-components makes it impossible as of now. And I think it'll cause issues for you too.

Say you have `/` and `/about`, and you arrive to `/about`, and then navigate to `/`. Both of these are dynamic in some way. For example here's what the Next.js build output gives me:

```js
Route (app)                                Size     First Load JS
┌ λ /                                      178 B          82.9 kB
├ λ /about                                 178 B          82.9 kB
└ ○ /favicon.ico                           0 B                0 B
+ First Load JS shared by all              77.1 kB
  ├ chunks/2443530c-d4a6bd7fc7f21e3c.js    50.5 kB
  ├ chunks/488-b4b91b085767513b.js         24.7 kB
  ├ chunks/main-app-100592265a5c8aaf.js    213 B
  └ chunks/webpack-f357de261c7e6f52.js     1.64 kB

Route (pages)                              Size     First Load JS
─ ○ /404                                   181 B          74.7 kB
+ First Load JS shared by all              74.5 kB
  ├ chunks/framework-8883d1e9be70c3da.js   45.1 kB
  ├ chunks/main-3f2d025144dc2872.js        27.6 kB
  ├ chunks/pages/_app-b555d5e1eab47959.js  195 B
  └ chunks/webpack-f357de261c7e6f52.js     1.64 kB

λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
○  (Static)  automatically rendered as static HTML (uses no initial props)
```

We see `/`, and `/about` are marked with `λ`, so they are dynamic, aka, rendered on demand. Well the issue is that there's a client side cache for server components, which as of next@13.4.4 stores the result, unlike GetServerSideProps, which would truly request new data on demand.

So my approach can't work because the component I use is cached client side. And so would be yours! (if you figure out the way)

https://github.com/vercel/next.js/assets/21013447/0ff4b75a-aff2-4a7d-bebf-a498d74a49c4

> My claim is that once you know which page you are serving, never mind rewrites, you can estimate which path you are rendering client side

You can see in the video, how in the client side navigation, the nav bar path, doesn't keep up with the URL, but refreshing it does show the right thing. 

I have a similar issue here https://next-js-react-gbg-git-feat-app-router-rsc-icyjoseph.vercel.app/collection:

https://github.com/vercel/next.js/assets/21013447/65ed5c62-adbe-49d7-ab2d-70500268731a

The list of Pokemon doesn't update, because it is stored in a cookie, which supposedly makes the route dynamic, but once the `collection` server component is cached client side... 

This works with the Pages Dir GSSP here: https://next-js-react-gbg.vercel.app/collection

> Of course, using `<a>` tags instead of `Link` to opt out from the client cache would be a way out, but in principle it goes against everything we are trying to achieve here, which is a seamless SPA like transition, and it'd look like this:


https://github.com/vercel/next.js/assets/21013447/6bd5c3b5-6fa8-4e19-8339-dd9beac39151

> No bueno I'd say

> Using some random query parameters might work, but it also feels hacky...
 
Anyway, this article can be helpful too https://www.inngest.com/blog/5-lessons-learned-from-taking-next-js-app-router-to-production#1-understand-the-two-caches

**Comment by didomenicom** [comment]
Thank you for the replies. I did a bit more searching and found [this in the Next 13 documentation](https://github.com/vercel/next.js/blob/canary/docs/02-app/02-api-reference/04-functions/use-router.mdx#router-events) that uses a client component in the root layout. Also, stated in [this document](https://nextjs.org/docs/app/api-reference/file-conventions/layout), `Because dashboard/layout.tsx doesn't re-render, the searchParams prop in the layout Server Component might become stale after navigation.`.

**Reply by icyJoseph** [reply]
I did it with intercepting routes though.
